datasafe-rest-impl/src/main/java/de/adorsys/datasafe/rest/impl/config/DatasafeConfig.java

```diff
@@ -1,39 +1,85 @@
 package de.adorsys.datasafe.rest.impl.config;
 
 import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.AmazonS3ClientBuilder;
 import de.adorsys.datasafe.business.impl.service.DaggerDefaultDatasafeServices;
 import de.adorsys.datasafe.business.impl.service.DaggerVersionedDatasafeServices;
 import de.adorsys.datasafe.business.impl.service.DefaultDatasafeServices;
 import de.adorsys.datasafe.business.impl.service.VersionedDatasafeServices;
 import de.adorsys.datasafe.directory.api.config.DFSConfig;
 import de.adorsys.datasafe.directory.impl.profile.config.DefaultDFSConfig;
 import de.adorsys.datasafe.storage.api.StorageService;
+import de.adorsys.datasafe.storage.impl.fs.FileSystemStorageService;
 import de.adorsys.datasafe.storage.impl.s3.S3StorageService;
-import de.adorsys.datasafe.types.api.resource.Uri;
+import lombok.extern.slf4j.Slf4j;
 import org.bouncycastle.jce.provider.BouncyCastleProvider;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
 
+import java.net.URI;
+import java.nio.file.FileSystems;
 import java.security.Security;
 import java.util.concurrent.Executors;
 
 /**
  * Configures default (non-versioned) Datasafe service that uses S3 client as storage provider.
  * Encryption provider: BouncyCastle.
  */
+@Slf4j
 @Configuration
 public class DatasafeConfig {
+    public final static String FILESYSTEM_ENV = "USE_FILESYSTEM";
```

Comment:
This can be simpler:
```
package de.adorsys.datasafe.rest.impl.config;

import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.AmazonS3ClientBuilder;
import de.adorsys.datasafe.business.impl.service.DaggerDefaultDatasafeServices;
import de.adorsys.datasafe.business.impl.service.DaggerVersionedDatasafeServices;
import de.adorsys.datasafe.business.impl.service.DefaultDatasafeServices;
import de.adorsys.datasafe.business.impl.service.VersionedDatasafeServices;
import de.adorsys.datasafe.directory.api.config.DFSConfig;
import de.adorsys.datasafe.directory.impl.profile.config.DefaultDFSConfig;
import de.adorsys.datasafe.storage.api.StorageService;
import de.adorsys.datasafe.storage.impl.fs.FileSystemStorageService;
import de.adorsys.datasafe.storage.impl.s3.S3StorageService;
import lombok.extern.slf4j.Slf4j;
import org.bouncycastle.jce.provider.BouncyCastleProvider;
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

import java.nio.file.Paths;
import java.security.Security;
import java.util.concurrent.Executors;

/**
 * Configures default (non-versioned) Datasafe service that uses S3 client as storage provider.
 * Encryption provider: BouncyCastle.
 */
@Slf4j
@Configuration
public class DatasafeConfig {
    public static final String FILESYSTEM_ENV = "USE_FILESYSTEM";

    @ConditionalOnProperty(FILESYSTEM_ENV)
    @Bean
    StorageService fsStorageService(DatasafeProperties properties) {
        String root = System.getenv(FILESYSTEM_ENV);
        log.info("==================== FILESYSTEM");
        log.info("build DFS to FILESYSTEM with root " + root);
        properties.setSystemRoot(root);
        return new FileSystemStorageService(Paths.get(root));
    }

    @ConditionalOnMissingBean(StorageService.class)
    @Bean
    StorageService s3StorageService(DatasafeProperties properties) {
        log.info("==================== AMAZONS3");
        AmazonS3 s3 = AmazonS3ClientBuilder.standard()
                .enablePathStyleAccess()
                .build();


        return new S3StorageService(
                s3,
                properties.getBucketName(),
                Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors())
        );
    }


    @Bean
    DFSConfig dfsConfig(DatasafeProperties properties) {
        return new DefaultDFSConfig(properties.getSystemRoot(), properties.getKeystorePassword());
    }

    /**
     * @return Default implementation of Datasafe services.
     */
    @Bean
    DefaultDatasafeServices datasafeService(StorageService storageService, DFSConfig dfsConfig) {

        Security.addProvider(new BouncyCastleProvider());

        return DaggerDefaultDatasafeServices
                .builder()
                .config(dfsConfig)
                .storage(storageService)
                .build();
    }

    @Bean
    VersionedDatasafeServices versionedDatasafeServices(StorageService storageService, DFSConfig dfsConfig) {

        Security.addProvider(new BouncyCastleProvider());

        return DaggerVersionedDatasafeServices
                .builder()
                .config(dfsConfig)
                .storage(storageService)
                .build();
    }
}
```

Reply:
locks more elegant but did not work (at least in my environment)

Reply:
`org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'documentController' defined in URL [jar:file:/usr/app/datasafe-rest-impl-*.jar!/BOOT-INF/classes!/de/adorsys/datasafe/rest/impl/controller/DocumentController.class]: Unsatisfied dependency expressed through constructor parameter 0; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'datasafeService' defined in class path resource [de/adorsys/datasafe/rest/impl/config/DatasafeConfig.class]: Unsatisfied dependency expressed through method 'datasafeService' parameter 1; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'dfsConfig' defined in class path resource [de/adorsys/datasafe/rest/impl/config/DatasafeConfig.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [de.adorsys.datasafe.directory.api.config.DFSConfig]: Factory method 'dfsConfig' threw exception; nested exception is java.lang.IllegalArgumentException: Resource location must be absolute BasePublicResource{uri=Uri{uri=/fuTD7AtO47n9dnVeA_xwXJaGbImmEMzxAICjeJl0Sls=/oXLO3K5HR0thXFTVEKXYSo3qMDLpWFh0MLQTU4vj8zM=/n8g39TmHSH6s_5PRt5R0b7Xx_qzI_HGv9T6F5icnWUk=/}}
        at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:769)
        at org.springframework.beans.factory.support.ConstructorResolver.autowireConstructor(ConstructorResolver.java:218)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.autowireConstructor(AbstractAutowireCapableBeanFactory.java:1341)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1187)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:555)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:515)
        at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:320)
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222)
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:318)
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:843)
        at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:877)
        at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:549)
        at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:142)
        at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:775)
        at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:397)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:316)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1260)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1248)
        at de.adorsys.datasafe.rest.impl.DatasafeRestApplication.main(DatasafeRestApplication.java:15)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48)
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:87)
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:50)
        at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:51)
Caused by: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'datasafeService' defined in class path resource [de/adorsys/datasafe/rest/impl/config/DatasafeConfig.class]: Unsatisfied dependency expressed through method 'datasafeService' parameter 1; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'dfsConfig' defined in class path resource [de/adorsys/datasafe/rest/impl/config/DatasafeConfig.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [de.adorsys.datasafe.directory.api.config.DFSConfig]: Factory method 'dfsConfig' threw exception; nested exception is java.lang.IllegalArgumentException: Resource location must be absolute BasePublicResource{uri=Uri{uri=/fuTD7AtO47n9dnVeA_xwXJaGbImmEMzxAICjeJl0Sls=/oXLO3K5HR0thXFTVEKXYSo3qMDLpWFh0MLQTU4vj8zM=/n8g39TmHSH6s_5PRt5R0b7Xx_qzI_HGv9T6F5icnWUk=/}}
        at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:769)
        at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:509)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1321)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1160)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:555)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:515)
        at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:320)
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222)
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:318)
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199)
        at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:277)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1248)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1168)
        at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:857)
        at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:760)
        ... 27 common frames omitted
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'dfsConfig' defined in class path resource [de/adorsys/datasafe/rest/impl/config/DatasafeConfig.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [de.adorsys.datasafe.directory.api.config.DFSConfig]: Factory method 'dfsConfig' threw exception; nested exception is java.lang.IllegalArgumentException: Resource location must be absolute BasePublicResource{uri=Uri{uri=/fuTD7AtO47n9dnVeA_xwXJaGbImmEMzxAICjeJl0Sls=/oXLO3K5HR0thXFTVEKXYSo3qMDLpWFh0MLQTU4vj8zM=/n8g39TmHSH6s_5PRt5R0b7Xx_qzI_HGv9T6F5icnWUk=/}}
        at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:627)
        at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:607)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1321)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1160)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:555)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:515)
        at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:320)
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222)
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:318)
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199)
        at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:277)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1248)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1168)
        at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:857)
        at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:760)
        ... 41 common frames omitted
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [de.adorsys.datasafe.directory.api.config.DFSConfig]: Factory method 'dfsConfig' threw exception; nested exception is java.lang.IllegalArgumentException: Resource location must be absolute BasePublicResource{uri=Uri{uri=/fuTD7AtO47n9dnVeA_xwXJaGbImmEMzxAICjeJl0Sls=/oXLO3K5HR0thXFTVEKXYSo3qMDLpWFh0MLQTU4vj8zM=/n8g39TmHSH6s_5PRt5R0b7Xx_qzI_HGv9T6F5icnWUk=/}}
        at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:185)
        at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:622)
        ... 55 common frames omitted
Caused by: java.lang.IllegalArgumentException: Resource location must be absolute BasePublicResource{uri=Uri{uri=/fuTD7AtO47n9dnVeA_xwXJaGbImmEMzxAICjeJl0Sls=/oXLO3K5HR0thXFTVEKXYSo3qMDLpWFh0MLQTU4vj8zM=/n8g39TmHSH6s_5PRt5R0b7Xx_qzI_HGv9T6F5icnWUk=/}}
        at de.adorsys.datasafe.types.api.resource.AbsoluteLocation.<init>(AbsoluteLocation.java:16)
        at de.adorsys.datasafe.directory.impl.profile.config.DefaultDFSConfig.dfsRoot(DefaultDFSConfig.java:131)
        at de.adorsys.datasafe.directory.impl.profile.config.DefaultDFSConfig.<init>(DefaultDFSConfig.java:74)
        at de.adorsys.datasafe.directory.impl.profile.config.DefaultDFSConfig.<init>(DefaultDFSConfig.java:60)
        at de.adorsys.datasafe.directory.impl.profile.config.DefaultDFSConfig.<init>(DefaultDFSConfig.java:42)
        at de.adorsys.datasafe.rest.impl.config.DatasafeConfig.dfsConfig(DatasafeConfig.java:63)
        at de.adorsys.datasafe.rest.impl.config.DatasafeConfig$$EnhancerBySpringCGLIB$$91d3d666.CGLIB$dfsConfig$2(<generated>)
        at de.adorsys.datasafe.rest.impl.config.DatasafeConfig$$EnhancerBySpringCGLIB$$91d3d666$$FastClassBySpringCGLIB$$cb8dc464.invoke(<generated>)
        at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:244)
        at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:363)
        at de.adorsys.datasafe.rest.impl.config.DatasafeConfig$$EnhancerBySpringCGLIB$$91d3d666.dfsConfig(<generated>)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154)
        ... 56 common frames omitted
08:50:09.148 [es-writer-1] WARN  es-error-logger - Failed to send events to Elasticsearch: aws-signing-proxy.aws-elasticsearch-proxy.svc.cluster.local
`

Reply:
This is because FILESYSTEM_ENV must be specified (with this approach) with URI-style:
file:///path/to/dir

Reply:
I.e. simply set env var:
`USE_FILESYSTEM=file:///tmp/datasafe`
And it will work

Reply:
yes, worked. thanks